geps/gep-1897.md

```diff
@@ -125,7 +125,231 @@ This GEP is the outcome of the TLS use cases #4 and #5 in
+mechanism for the cluster operator gateway role to override gateway to backend TLS settings is not covered here, but can
+be addressed in a future update should the need arise.
+
+TLSConnectionPolicy is defined as a Direct Policy Attachment without defaults or overrides, applied to a Service that
+accesses the backend in question, where the TLSConnectionPolicy resides in the same namespace as the Service it is
+applied to.  The TLSConnectionPolicy and the Service must reside in the same namespace in order to prevent the
+complications involved with sharing trust across namespace boundaries.  By choosing the Service resource rather than the
+Route resource, we can reuse the same TLSConnectionPolicy for all the different Routes that might point to this Service.
+For the use case where certificates are stored in their own namespace, users may create Secrets and use ReferenceGrants
+for a TLSConnectionPolicy-to-Secret binding.
+
+In the API defined here, the definition of TrustedCACertRefs follows a convention established by TLSRoute in
+https://github.com/kubernetes-sigs/gateway-api/blob/18e79909f7310aafc625ba7c862dfcc67b385250/apis/v1beta1/gateway_types.go#L340
+
+One of the areas of concern for this API is that we need to indicate how and when the API implementations should use the
+backend destination certificate authority.  This solution proposes, as introduced in
+[GEP-713](https://gateway-api.sigs.k8s.io/geps/gep-713/), that the implementation
+should watch the connections to a specific port on a specified targetRef (such as a Service), and if the port and
+Service match a TLSConnectionPolicy, then assume the connection is TLS, and verify that the targetRef’s certificate can
+be validated by the provided trusted CA certificates before the connection is made.  On the question of how to signal
+that there was a failure in the certificate validation, this is left up to the implementation to return a response error
+that is appropriate, such as one of the HTTP error codes: 400 (Bad Request), 401 (Unauthorized), 403 (Forbidden), or
+other signal that makes the failure sufficiently clear to the requester without revealing too much about the transaction,
+based on established security requirements.
+
+Not covered here, but possible to add would be additional configuration options mentioned in [SIG-NET Gateway API: TLS
+to the K8s.Service/Backend](https://docs.google.com/document/d/1RTYh2brg_vLX9o3pTcrWxtZSsf8Y5NQvIG52lpFcZlo).  All of
+these are currently implementation-dependent, with the following recommended defaults:
+- Server Name Indication: enables passing of the server name through server name indication, in the TLS transaction, to
+assist with selection of certificates when several hosts share the same IP address. (default to enabled)
+- Subject Alternative Name certificates: enable the use of a single certificate that can serve multiple domains.
```

Comment:
What is "enabled"? this is a property of the verification. So does enabled mean we should verify it (we ought to by default or it's the equivalent of -k in curl).

If so what are we verifying? It's not just a bool.

Browsers handle this by asserting the DNS name matches the certificate, but a Kubernetes service doesn't really have as strict of a mapping.

Reply:
I think what "enabled" means is that a SAN (multi-domain) certificate will be accepted when true. When false, a certificate containing SANs will be rejected right away. The verification is always done (unless disabled otherwise with a different setting).

Regarding what we are verifying it's a good question: in Kubernetes (or service meshes like Istio) certificates may contain SPIFFE ids but the client needs to be provided with a mapping (service to valid SPIFFE id) to authorize the server.

Reply:
SAN is the only non-deprecated way to encode a DNS name as far as I know (CN was deprecated 20 years ago). So it's not a question of if we verify San or not - we MUST.

The question is *what* SAN do we verify. Tou don't have to get into spife even, the pod could have any certificate..

Reply:
the simplest case would be to assume svc.ns.svc.cluster.local but I am not sure that is sufficient, or even a common usage

Reply:
I agree that SAN certs have been around for long and CN-only certs have been deprecated so this setting is not needed/useful.

Regarding "what SAN do we verify": may be that needs to be an additional field here with defaults for inferred names like "svc.ns.svc.cluster.local"

Reply:
My intention for "default is enabled" is that the option to use each of these is on by default.  Implementations could choose whether to allow any option to be disabled.  If that doesn't apply to SAN certificates, and you expect that no one would ever want to program "no SAN certificates allowed", then I can remove this and add a section that SAN certificates are always allowed.

Reply:
"Using SAN" is not a meaningful statement.. it's not our choice,it's the TLS spec. We need to specify *which* SAN is allowed. This is absolutely a hard requirement for anything doing TLS

Reply:
Your feedback is not clear to me, can you help me out?  What do you expect to need to do with the SAN values other than provide them to the target?  Why would you choose only one?

Reply:
I am not sure there is clarity here on what SAN is.

We have a client (the gateway) and server (the application). We want to use TLS to provide authentication, encryption, and integrity. SAN is what provides authentication. 

The client connects to the server and the server presents a certificate. The certificate has a SAN (usually - some only use CN but we mentioned it was deprecated 25 years ago so for all intents and purposes we can assume SAN is present to keep things simple). For a real world example, when you connect to google.com the SAN would be google.com.

Now the client wants to verify the server is who they expect (this is the authentication aspect of TLS; without this we only have encryption). First they verify the CA that signed the certificate is trusted (otherwise the data is not trusted). Then they verify who they are (the SAN) is aligned with their expectations.

In a browser scenario the expectation is always the DNS name matches the SAN.

In gateway, I don't think there is an obvious implicit expectation. We can default to the .cluster.local name perhaps but that is not going to be sufficient in all cases

Reply:
I as an end-user would like to provide a list of host names here (SAN) for Upstream TLS verification

xref: https://github.com/projectcontour/contour/issues/5520

Reply:
@howardjohn I suppose if I just delete SAN as an implementation-dependent option, that would close this issue. Sound good?

Reply:
Presumably, implementations are verifying _something_ today right? Since this GEP is going in as provisional, I think SAN verification can remain implementation-specific for now; however, I do agree that there needs to be a SAN list before this moves to implementable

Reply:
SAN validation is a very very hard requirement. Without it we are basically doing `--insecure-skip-verify`

Reply:
> In a browser scenario the expectation is always the DNS name matches the SAN.
> 
> In gateway, I don't think there is an obvious implicit expectation. We can default to the .cluster.local name perhaps but that is not going to be sufficient in all cases

Here we're dealing only with the connection from gateway to backend.  I expect that one of the hostnames on the HTTPRoute matches the SAN.  Is that too much to expect?

Reply:
The HTTPRoute hostname is the public internet facing hostname, not the internal service. So I'm a route from example.com->example-svc, I would Not expect the workload to have a cert for example.com

it could be example-svc.ns.svc.cluster.local maybe - but there is no standard

Reply:
Isn't that very service-mesh specific to not expect the pod to have a cert for example.com?  Alternatively, with SANs, it could list example.com and example-svc.ns.svc.cluster.local.

Reply:
It is not related to service mesh in any way. I wouldn't expect them to have a cert for example.com since the Gateway itself has that. And it's not clear why you would serve the same cert in 2 places?

Reply:
To be fair they could have any cert - even google.com (with their own ca, of course). So anything is *possible*

Reply:
The internal workload on the pod can serve requests either from a Gateway or another internal workload (the mesh scenario) and in both cases is likely to serve the same certificate (unless SNI is used?) which is more likely to be example-svc and not example.com.